### ISSUE TITLE
08-k-nearest-neighbours-winequality-red.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/abgonzalez-introduction-machine-learning#4]
@abgonzalez 

## Fortalezas (Muy buenas)

**1. EDA profesional y completo**
**2. Análisis de múltiples métricas (no solo accuracy)**
**3. Uso avanzado de SMOTE**
**4. Comparación de distancias (Euclidean vs Manhattan)**
**5. Barrido de k con análisis detallado**
**6. Guardado profesional**
**7. Función de predicción avanzada**

---

### 1. **Modelo final no es el modelo optimizado**

**Problema:**
```python
# El LOOP se entrena con SMOTE y ESCALADO:
for k in range(1, 21):
    modelo = KNeighborsClassifier(n_neighbors=k)
    modelo.fit(X_train_resampled, y_train_resampled)  # ← CON SMOTE
    pred = modelo.predict(X_test_scaled)

# Pero el modelo FINAL se entrena sin SMOTE:
modelo_knn_elegido = KNeighborsClassifier(n_neighbors=2) 
modelo_knn_elegido.fit(X_train, y_train)  # ← SIN SMOTE, SIN ESCALAR 
```
**¿Qué ha pasado?**
1. Optimizaste bien con `X_train_resampled` (balanceado)
2. Encontraste que k=2 es mejor
3. Pero luego entrenaste con `X_train` (desbalanceado, sin escalar)

**Cómo corregir:**
```python
# CORRECTO:
modelo_knn_elegido = KNeighborsClassifier(n_neighbors=2)
modelo_knn_elegido.fit(X_train_resampled, y_train_resampled)  # Igual al loop
```

---

### 2. **Función de predicción escala datos pero modelo no fue escalado**

```python
def predecir_calidad_vino_optimizado(valores):
    # ...
    muestra_escalada = scaler.transform(muestra)  # ← Escala
    prediccion = modelo_knn_elegido.predict(muestra_escalada)  # ← Sobre modelo SIN escalar
```

**Problema:**
- El modelo se entrenó con `X_train` (sin escalar)
- Pero la función predice sobre `muestra_escalada` (escalada)
- **Data mismatch** → predicciones incorrectas

Podrías usar el scaler guardado:
```python
modelo_knn_elegido = KNeighborsClassifier(n_neighbors=2)
modelo_knn_elegido.fit(X_train_resampled, y_train_resampled)

# En predicción:
def predecir_calidad_vino_optimizado(valores):
    muestra = pd.DataFrame([valores], columns=X.columns)
    muestra_escalada = scaler.transform(muestra)
    prediccion = modelo_knn_elegido.predict(muestra_escalada)[0]
    # ...
```

---

### 3. **Sugiero para este ejercicio usar `stratify=y` en train_test_split**

```python
# AHORA (sin stratify):
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# SERÍA:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)
```

**¿Por qué importa?**
Con `stratify=y`, garantizas que train y test tienen la misma distribución de clases. Especialmente importante con clases desbalanceadas.

---

Notebook de muy alto nivel técnico pero con algún **fallo de consistencia** que impacta en la reproducibilidad. Enhorabuena, corrige estos temas y lo tendrás de 10.

**Comment by abgonzalez** [comment]
Gracias Bea, ya he hecho los cambios. La verdad es que probe con SMOTE y sin el SMOTE y en un momento me parecio que no daban buenos resultados. Y al final creo que me complique yo sola. 

Gracias, si ves cualquier otra cosa, dimelo
